algorithms/mopt/mopt_optimizer.cpp

```diff
@@ -0,0 +1,180 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+
+#include "fuzzuf/algorithms/mopt/mopt_optimizer.hpp"
+
+#include <random>
+#include <vector>
+
+#include "fuzzuf/mutator/havoc_case.hpp"
+#include "fuzzuf/optimizer/keys.hpp"
+#include "fuzzuf/utils/common.hpp"
+
+namespace fuzzuf::optimizer {
+
+using fuzzuf::mutator::INSERT_AEXTRA;
+using fuzzuf::mutator::INSERT_EXTRA;
+using fuzzuf::mutator::OVERWRITE_WITH_AEXTRA;
+using fuzzuf::mutator::OVERWRITE_WITH_EXTRA;
+
+MOptParticle::MOptParticle() {}
+MOptParticle::~MOptParticle() {}
+
+MOptOptimizer::MOptOptimizer() {
+  fuzzuf::optimizer::Store::GetInstance().InitKey(
+      fuzzuf::optimizer::keys::NewTestcases, (u64)0);
+  fuzzuf::optimizer::Store::GetInstance().InitKey(
+      fuzzuf::optimizer::keys::HavocOperatorFinds);
+  fuzzuf::optimizer::Store::GetInstance().InitKey(
+      fuzzuf::optimizer::keys::SelectedCaseHistogram);
+  fuzzuf::optimizer::Store::GetInstance().InitKey(
+      fuzzuf::optimizer::keys::LastSpliceCycle, (u32)0);
+
+  for (auto& p : swarm) {
+    for (auto& pos : p.position)
+      pos = fuzzuf::utils::random::Random<double>(min_position, max_position);
+    p.velocity.fill(0);
+  }
+
+  idx = 0;
+  time = 0;
+
+  UpdateInertia();
+}
+
+MOptOptimizer::~MOptOptimizer() {}
+
+void MOptOptimizer::SetScore(size_t i, double score) {
+  auto& p = swarm[idx];
+  p.fitness[i] = score;
+}
+
+void MOptOptimizer::UpdateLocalBest() {
+  auto& p = swarm[idx];
+
+  if (unlikely(time == 0)) {
+    for (size_t i = 0; i < p.fitness.size(); i++) {
+      p.best_fitness[i] = p.fitness[i];
+      p.best_position[i] = p.position[i];
+    }
+  }
+
+  for (size_t i = 0; i < p.fitness.size(); i++) {
+    if (p.fitness[i] > p.best_fitness[i]) {
+      p.best_fitness[i] = p.fitness[i];
+      p.best_position[i] = p.position[i];
+    }
+  }
+}
+
+void MOptOptimizer::UpdateGlobalBest() {
+  if (unlikely(time == 0)) {
+    // TODO?
+  }
+
+  std::array<u64, NUM_CASE> havoc_operator_dist;
+  havoc_operator_dist.fill(0);
+
+  for (size_t i = 0; i < havoc_operator_finds[0].size(); i++) {
+    havoc_operator_dist[i] =
+        havoc_operator_finds[0][i] + havoc_operator_finds[1][i];
+  }
+
+  std::discrete_distribution<u32> dist(havoc_operator_dist.begin(),
+                                       havoc_operator_dist.end());
+  std::vector<double> prob = dist.probabilities();
+
+  for (size_t i = 0; i < prob.size(); i++) {
+    best_position[i] = prob[i];
+  }
+}
+
+void MOptOptimizer::UpdateInertia() {
+  w = (W_INIT - W_END) * (G_MAX - g_now) / G_MAX + W_END;
+  ++g_now %= G_MAX;
+}
+
+bool MOptOptimizer::IncrementSwarmIdx() {
+  idx++;
+  idx %= swarm.size();
+  return idx == 0;
```

Comment:
The name `IncrementSwarmIdx` does not suggest that this functions returns true if idx did a lap and becomes 0. But I don't have a good idea to resolve it. Renaming it to `IncrementSwarmIdxAndCheckIfMadeRound` might be slightly better...
But any ideas?

Reply:
fixed in baf38611e98a2c7d89aa9bf6995a83c684de32a5

rename `IncrementSwarmIdx` to `NextSwarmIdx` and made it return incremented index

Reply:
That's smart.
But I think you forgot to replace the type of return values.
https://github.com/fuzzuf/fuzzuf/commit/baf38611e98a2c7d89aa9bf6995a83c684de32a5#diff-6a8c4972fbfb3703282f4f41c9b8a685401f29bc8c395d701a71bb15fdf0d496R118

Reply:
fixed in 91e3ca33c40da44d60aa47847b59a2c85d5c1dd0